### ISSUE TITLE
Node-RED

### ISSUE THREAD

**Issue by aaronpk** [bitfocus/companion-module-requests#217]
Node-RED is a visual programming environment with support for a lot of devices and integrations itself. 

My primary use of Node-RED is as a home automation controller, creating rules to turn on and off lights. I'd like to be able to hook up buttons on my Stream Deck to control the lights in my studio along with using it to control an ATEM.

If Companion could support Node-RED then I could write Node-Red rules to turn on lights in response to pressing a button in Companion. 

I am not totally sure how this integration would work, but I suspect the way to do it is to [create a new "node" type](https://nodered.org/docs/creating-nodes/) in Node-RED. This is how Node-RED's integration with Home Assistant works already.

**Comment by zmcandee** [comment]
I have zero node-RED experience so take this with a grain of salt but I believe both node-RED and Companion have MQTT modules which you could use to send button presses to Node-RED.

**Comment by aaronpk** [comment]
clever, I will look into that!

**Comment by herzogmedia** [comment]
yes, I'd rather not make new node-types but use existing control protocols.
Instead of mqtt, you could also use [osc-nodes](https://flows.nodered.org/node/node-red-contrib-osc) and send OSC-Commands from companion.

**Comment by peppertechie** [comment]
I have Node-Red running on a Pi and using it to run commands and other tasks on the Pi from Companion.  I've used both the UDP and TCP input nodes successfully.  I'm not sure there would be a one size fits all Node Red node/Companion module pair.  
What the generic modules are missing is feedback, but I don't think that necessarily fits into the spirit of Companion, but would make integrating with node-RED more powerful.

**Comment by josephdadams** [comment]
Closing as **not planned**.

This request is labelled as having no external control protocol (or as unlikely to be feasible) — the device or software doesn't expose an interface Companion could talk to, so a module can't be built for it as things stand.

If the manufacturer adds a control API later, please comment and we'll happily reopen.
